### PR TITLE
Unified analytics events migration for `BuildEvent`

### PR DESCRIPTION
**Pull request by eliasyishak**
Related to tracker issue:
- https://github.com/flutter/flutter/issues/128251

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/pubspec.yaml [hunk 1]
```diff
@@ -48,18 +48,18 @@ dependencies:
   http_multi_server: 3.2.1
   convert: 3.1.1
   async: 2.11.0
-  unified_analytics: 5.0.0
```
Comment (eliasyishak) [inline packages/flutter_tools/pubspec.yaml:51]: @christopherfujino @andrewkolos this version update will likely take some time to make it into google3 (git -> gerrit -> google3).. so i'm expecting Google testing to fail for a day or 2

Comment (christopherfujino) [reply packages/flutter_tools/pubspec.yaml:51]: Do we have the option to manually roll it? Not necessarily for this change, but to lift the constraint that EACH PR you land in dart-lang/tools, the corresponding tool PR must wait for rolls.

Comment (eliasyishak) [reply packages/flutter_tools/pubspec.yaml:51]: I am able to run git -> gerrit by manually opening up a CL pretty easily. This job only runs twice a week so being able to manually do it is ideal.

But for the google3 roll, this job runs daily and I don't know how to do those manually yet

Comment (christopherfujino) [reply packages/flutter_tools/pubspec.yaml:51]: Ok, daily is not too bad. Why don't we plan to roll manually to gerrit (btw, where is the gerrit repo?)

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 2]
```diff
@@ -581,17 +582,34 @@ return result.exitCode != 0 &&
...
-      type: 'ios',
-      command: xcodeBuildExecution.buildCommands.toString(),
-      settings: xcodeBuildExecution.buildSettings.toString(),
+
+    const String label = 'xcode-bitcode-failure';
+    const String buildType = 'ios';
+    final String command = xcodeBuildExecution.buildCommands.toString();
+    final String settings = xcodeBuildExecution.buildSettings.toString();
+
+    BuildEvent(label,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/ios/mac.dart:601]: nit
```suggestion
    BuildEvent(
      label,
```

File: packages/flutter_tools/lib/src/project.dart [hunk 3]
```diff
@@ -761,8 +762,18 @@ The detected reason was:
 ''');
     if (deprecationBehavior == DeprecationBehavior.ignore) {
       BuildEvent('deprecated-v1-android-embedding-ignored', type: 'gradle', flutterUsage: globals.flutterUsage).send();
+      globals.analytics.send(Event.flutterBuildInfo(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/project.dart:765]: nit
```suggestion
      globals.analytics.send(
        Event.flutterBuildInfo(
```

File: packages/flutter_tools/lib/src/project.dart [hunk 4]
```diff
@@ -761,8 +762,18 @@ The detected reason was:
...
+        label: 'deprecated-v1-android-embedding-ignored',
+        buildType: 'gradle',
+      ));
+
     } else { // DeprecationBehavior.exit
       BuildEvent('deprecated-v1-android-embedding-failed', type: 'gradle', flutterUsage: globals.flutterUsage).send();
+      globals.analytics.send(Event.flutterBuildInfo(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/project.dart:772]: nit

```suggestion
      globals.analytics.send(
        Event.flutterBuildInfo(
```

File: packages/flutter_tools/lib/src/web/compile.dart [hunk 5]
```diff
@@ -127,15 +131,23 @@ class WebBuilder {
     } finally {
       status.stop();
     }
+
+    final String buildSettingsString = _buildEventAnalyticsSettings(
+        config: compilerConfig,
+        buildInfo: buildInfo,
+      );
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/web/compile.dart:138]: nit

```suggestion
      config: compilerConfig,
      buildInfo: buildInfo,
    );
```

File: packages/flutter_tools/test/commands.shard/hermetic/build_ios_test.dart [hunk 6]
```diff
@@ -587,6 +594,13 @@ void main() {
           parameters:CustomDimensions(),
         ),
       ));
+
+      expect(fakeAnalytics.sentEvents, contains(
+        Event.flutterBuildInfo(
+            label: 'plist-impeller-enabled',
+            buildType: 'ios',
+          )
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/build_ios_test.dart:602]: nit
```suggestion
          label: 'plist-impeller-enabled',
          buildType: 'ios',
        )
```

File: packages/flutter_tools/test/commands.shard/hermetic/build_ios_test.dart [hunk 7]
```diff
@@ -642,6 +657,13 @@ void main() {
           parameters:CustomDimensions(),
         ),
       ));
+
+      expect(fakeAnalytics.sentEvents, contains(
+        Event.flutterBuildInfo(
+            label: 'plist-impeller-disabled',
+            buildType: 'ios',
+          )
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/build_ios_test.dart:665]: nit

```suggestion
          label: 'plist-impeller-disabled',
          buildType: 'ios',
        )
```

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 8]
```diff
@@ -307,6 +325,15 @@ void main() {
...
         ),
       ));
+      expect(testUsage.events, hasLength(4));
+
+      expect(fakeAnalytics.sentEvents, hasLength(4));
+      expect(fakeAnalytics.sentEvents, contains(
+        Event.flutterBuildInfo(
+            label: 'gradle-random-event-label-failure',
+            buildType: 'gradle',
+          )
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:335]: nit, de-dent and trailing comma

```suggestion
          label: 'gradle-random-event-label-failure',
          buildType: 'gradle',
        ),
```

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 9]
```diff
@@ -400,6 +428,16 @@ void main() {
...
         ),
       ));
+      expect(testUsage.events, hasLength(2));
+
+      expect(fakeAnalytics.sentEvents, hasLength(2));
+      expect(fakeAnalytics.sentEvents, contains(
+        Event.flutterBuildInfo(
+            label: 'gradle-random-event-label-failure',
+            buildType: 'gradle',
+          )
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:438]:
```suggestion
          label: 'gradle-random-event-label-failure',
          buildType: 'gradle',
        ),
```

File: packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart [hunk 10]
```diff
@@ -133,6 +141,14 @@ void main() {
...
         ),
       ));
+      expect(testUsage.events, hasLength(2));
+
+      expect(
+          fakeAnalytics.sentEvents,
+          unorderedEquals(<Event>[
+            Event.flutterBuildInfo(label: 'app-not-using-android-x', buildType: 'gradle'),
+            Event.flutterBuildInfo(label: 'gradle-random-event-label-failure', buildType: 'gradle'),
+          ]));
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/android_gradle_builder_test.dart:151]:
```suggestion
        fakeAnalytics.sentEvents,
        unorderedEquals(<Event>[
          Event.flutterBuildInfo(label: 'app-not-using-android-x', buildType: 'gradle'),
          Event.flutterBuildInfo(label: 'gradle-random-event-label-failure', buildType: 'gradle'),
        ]),
      );
```

File: packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart [hunk 11]
```diff
@@ -400,6 +435,13 @@ void main() {
         }),
       ),
     ));
+    expect(fakeAnalytics.sentEvents, hasLength(1));
+    expect(
+        fakeAnalytics.sentEvents,
+        contains(Event.flutterBuildInfo(
+            label: 'gradle-expected-file-not-found',
+            buildType: 'gradle',
+            settings: 'androidGradlePluginVersion: 7.5, fileExtension: .aab')));
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/gradle_find_bundle_test.dart:444]:
```suggestion
    expect(
      fakeAnalytics.sentEvents,
      contains(
        Event.flutterBuildInfo(
          label: 'gradle-expected-file-not-found',
          buildType: 'gradle',
          settings: 'androidGradlePluginVersion: 7.5, fileExtension: .aab',
        ),
      ),
    );
```

File: packages/flutter_tools/test/general.shard/web/compile_web_test.dart [hunk 12]
```diff
@@ -103,6 +109,14 @@ void main() {
       ),
     );
 
+    expect(fakeAnalytics.sentEvents, unorderedEquals(<Event>[
+      Event.flutterBuildInfo(
+            label: 'web-compile',
+            buildType: 'web',
+            settings: 'RunWasmOpt: none; WasmOmitTypeChecks: false; wasm-compile: true; web-renderer: auto;',
+          ),
+    ]));
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/web/compile_web_test.dart:118]:
```suggestion
    expect(
      fakeAnalytics.sentEvents,
      unorderedEquals(<Event>[
        Event.flutterBuildInfo(
          label: 'web-compile',
          buildType: 'web',
          settings: 'RunWasmOpt: none; WasmOmitTypeChecks: false; wasm-compile: true; web-renderer: auto;',
        ),
      ],
    ),
  );
```

**Review by christopherfujino — APPROVED** [review]
Other than formatting nits, this LGTM

File: packages/flutter_tools/pubspec.yaml (hunk 1 above)
Comment (eliasyishak) [reply packages/flutter_tools/pubspec.yaml]: I do plan on rolling gerrit manually, and the changes are made in the `dart-lang/sdk` repo but it goes through Gerrit when you want to contribute to the sdk, here is the [CL](https://dart-review.googlesource.com/c/sdk/+/333061) for making the version bump

Their [contributing guide](https://github.com/dart-lang/sdk/blob/main/CONTRIBUTING.md#uploading-the-patch-for-review) also explains how to get your CLs submitted to Gerrit

**Comment by christopherfujino** [comment]
@eliasyishak I think the framework widget tests failure is due to test ordering, and was "fixed" upstream (by adding the no-shuffle annotation)

File: packages/flutter_tools/pubspec.yaml (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/pubspec.yaml]: oh, if that's the Dart SDK, then that's probably only for the analyzer and dart-dev, right? The copybara script pulls from dart-lang/tools, right (I'm actually asking, I don't really know)? We don't need to manually update the dart sdk unless that's blocking the copybara github -> internal google roll.

**Pull request merged**